### ISSUE TITLE
New scanned paged not displayed when window is not visible

### ISSUE THREAD

**Issue by tiramiseb** [jflesch/paperwork#515]
I'm experiencing a weird problem...

As long as the Paperwork window is visible (or as long as I am displaying the GNOME virtual desktop where Paperwork is placed), everything works perfectly.

But if I switch to a different workspace/virtual desktop when scanning, the Paperwork window doesn't change (however, the scanned page is correctly stored in the filesystem).

Afterwards :
* for a new document, switching to an older document and back to "new document" seems to end the display process, creating a new entry in the documents list
* for new pages on an existing document, switching to another document and back to the previous one works too

By the way, for a new document, after having the aforementioned behaviour, if I :
* click on another document
* then click on "new document" (as described before)
* then wait for the new entry in the documents list to be created (it takes approx. 1 second)
* then switch to another workspace BEFORE the page is rendered
... then, when switching back to Paperwork, the enlarged thumbnail is displayed, not the "correct quality" image for the document

**Comment by tiramiseb** [comment]
When the Paperwork window is hidden behind a maximized window on the same workspace, the problem does not occur.

**Comment by jflesch** [comment]
Looks like a Cairo rendering bug.
Can you remind me which GNU/Linux distribution do you use ? And which Gnome version do you have ?

**Comment by jflesch** [comment]
Ping ?

**Comment by tiramiseb** [comment]
I'm using Ubuntu GNOME 16.04, with GNOME Shell 3.18.

**Comment by jflesch** [comment]
Actually, I reproduced on Debian sid too.

I'm unsure what is happening here. No obvious bug / trace in Paperwork output.

However, from what I can see, this is not a big issue. There is pretty much no point where the user is stuck. So for now, I'm going to delay this ticket.

**Comment by tiramiseb** [comment]
I am not stuck, but it makes it harder to use Paperwork while doing something else...

**Comment by tiramiseb** [comment]
I have switched to KDE this weekend.

On KDE, this bug does not appear when switching to another activity while scanning.

**Comment by jflesch** [comment]
Maybe related to #522

**Comment by tiramiseb** [comment]
I don't think so because #522 has no problem when adding pages to an existing documents.

**Comment by jflesch** [comment]
Yep, but in both cases, I suspect a Gtk/Cairo rendering issue (or worst, a main loop issue) :)

**Comment by jflesch** [comment]
Anyway, we will have to check if 1.1.0 doesn't already fix the issue.

Prior to 1.1.0, I didn't call Gdk.threads_init() because it caused issues with Gtk dialogs in Paperwork. However I am clearly using threads (and a lot). Since I've solved the issues with the dialogs, I am now able to call Gdk.threads_init(). Hopefully, this is what caused these main loop issues and/or rendering problems.

If not, I'll have to do some nasty debugging to find the root causes :(

**Comment by jflesch** [comment]
Update: Gdk.threads_init() is actually deprecated (--> wasn't included in Paperwork 1.1).

I've also installed Ubuntu Gnome 16.04 on my desktop because it's currently the only distribution that supports correctly my graphic card with Steam ^^. Anyway, I've ended up with a bunch of weird rendering bugs in Paperwork I didn't have with Debian (stable or sid). I'm more and more suspecting bug(s) in the Glib/Cairo/Gdk/Gtk/Gnome-shell versions of Ubuntu Gnome 16.04. And I am unable to spot where they come from yet ... :/
